plugins/search-backend-module-pg/src/PgSearchEngine/PgSearchEngine.ts

```diff
@@ -26,28 +26,112 @@ import {
   DatabaseStore,
   PgSearchQuery,
 } from '../database';
+import { v4 as uuid } from 'uuid';
+import { Config } from '@backstage/config';
 
+/**
+ * Search query that the Postgres search engine understands.
+ * @public
+ */
 export type ConcretePgSearchQuery = {
   pgQuery: PgSearchQuery;
   pageSize: number;
 };
 
+/**
+ * Options available for the Postgres specific query translator.
+ * @public
+ */
+export type PgSearchQueryTranslatorOptions = {
+  highlightOptions: PgSearchHighlightOptions;
+};
+
+/**
+ * Postgres specific query translator.
+ * @public
+ */
+export type PgSearchQueryTranslator = (
+  query: SearchQuery,
+  options: PgSearchQueryTranslatorOptions,
+) => ConcretePgSearchQuery;
+
+/**
+ * Options to instantiate PgSearchEngine
+ * @public
+ */
+export type PgSearchOptions = {
+  config: Config;
+  database: PluginDatabaseManager;
+};
+
+/**
+ * Options for highlighting search terms
+ * @public
+ */
+export type PgSearchHighlightOptions = {
+  useHighlight?: boolean;
+  maxWords?: number;
+  minWords?: number;
+  shortWord?: number;
+  highlightAll?: boolean;
+  maxFragments?: number;
+  fragmentDelimiter?: string;
+  preTag: string;
+  postTag: string;
+};
+
 export class PgSearchEngine implements SearchEngine {
-  constructor(private readonly databaseStore: DatabaseStore) {}
+  private readonly highlightOptions: PgSearchHighlightOptions;
+  constructor(private readonly databaseStore: DatabaseStore, config: Config) {
+    const uuidTag = uuid();
+    const highlightConfig = config.getOptionalConfig(
+      'search.pg.highlightOptions',
+    );
+
+    const highlightOptions: PgSearchHighlightOptions = {
+      preTag: `<${uuidTag}>`,
+      postTag: `</${uuidTag}>`,
+      useHighlight: highlightConfig?.getOptionalBoolean('useHighlight') ?? true,
+      maxWords: highlightConfig?.getOptionalNumber('maxWords') ?? 35,
+      minWords: highlightConfig?.getOptionalNumber('minWords') ?? 15,
+      shortWord: highlightConfig?.getOptionalNumber('shortWord') ?? 3,
+      highlightAll:
+        highlightConfig?.getOptionalBoolean('highlightAll') ?? false,
+      maxFragments: highlightConfig?.getOptionalNumber('maxFragments') ?? 0,
+      fragmentDelimiter:
+        highlightConfig?.getOptionalString('fragmentDelimiter') ?? ' ... ',
+    };
+    this.highlightOptions = highlightOptions;
+  }
 
+  /**
+   * @deprecated This will be removed in a future release, please us fromConfig instead
+   */
   static async from(options: {
     database: PluginDatabaseManager;
+    config: Config;
   }): Promise<PgSearchEngine> {
     return new PgSearchEngine(
       await DatabaseDocumentStore.create(await options.database.getClient()),
+      options.config,
+    );
+  }
+
+  static async fromConfig({ config, database }: PgSearchOptions) {
```

Comment:
Hey!  Thanks for splitting this out.  The pattern for these is more typically `fromConfig(config, options)`, where `config` is its own argument in the first position, and the next argument is an object containing other dependencies.  Would you mind updating?

Reply:
I've cleaned this up, thanks